File: platform/cli-go/internal/commands/feature.go
package commands

import (
	"fmt"
	"strings"
	"time"

	"github.com/elloloop/devpod/platform/cli-go/internal/format"
	"github.com/elloloop/devpod/platform/cli-go/internal/git"
	"github.com/elloloop/devpod/platform/cli-go/internal/workspace"
	"github.com/spf13/cobra"
)

func startWork(name string, changeType workspace.ChangeType) error {
	// Check for uncommitted changes -- refuse if dirty
	if workspace.HasUncommittedChanges() {
		return fmt.Errorf("%s\n\n  Save your changes first with: devpod diff\n  Or stash them with: git stash",
			format.ErrorMsg("You have unsaved changes."))
	}

	// Check for pending rebase -- refuse with instructions
	if err := workspace.CheckPendingRebase(); err != nil {
		return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
	}

	config := workspace.LoadConfig()
	slug := workspace.Slugify(name)

	// Reject names containing ~ (reserved for versions branches)
	if strings.Contains(name, "~") {
		return fmt.Errorf("%s\n\n  The ~ character is reserved. Choose a different name.",
			format.ErrorMsg("Invalid feature name."))
	}

	// Check for duplicates
	existing, _ := workspace.LoadFeature(slug)
	if existing != nil {
		return fmt.Errorf("%s\n\n  A feature named \"%s\" already exists.\n  Pick a different name, or switch to it: devpod switch \"%s\"",
			format.ErrorMsg("Duplicate feature name."), name, name)
	}

	prefix := string(changeType)
	if changeType == "unknown" {
		prefix = "feature"
	}
	branch := prefix + "/" + slug
	versionsBranch := workspace.VersionsBranchName(branch)

	// Fetch latest and create branch from origin/main
	_ = git.FetchMain()
	if err := git.CreateBranch(branch, "origin/"+config.DefaultBranch); err != nil {
		return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
	}

	// Create feature data
	feature := workspace.FeatureData{
		Name:           name,
		Type:           changeType,
		Slug:           slug,
		Branch:         branch,
		VersionsBranch: versionsBranch,
		Created:        time.Now().UTC().Format(time.RFC3339),
		Diffs:          []string{},
		Status:         "active",
		SnapshotCount:  0,
	}

	if err := workspace.EnsureDevpodDir(); err != nil {
		return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
	}

	// Create the versions branch (orphan, seeded from current state)
	if err := workspace.EnsureVersionsBranch(feature); err != nil {
		// Non-fatal: versions branch creation may fail in edge cases
		fmt.Println(format.DimText("  Note: could not create versions branch. It will be created on first diff."))
	}

	if err := workspace.SaveFeature(feature); err != nil {
		return fmt.Errorf("%s", format.ErrorMsg(err.Error()))
	}

	fmt.Printf("%s Started %s: %s\n", format.SuccessMsg("\u2713"), format.FeatureTypePrefix(string(changeType)), name)
	fmt.Println(format.DimText(fmt.Sprintf("  Branch: %s", branch)))
	fmt.Println(format.DimText(fmt.Sprintf("  Versions: %s", versionsBranch)))
	fmt.Println(format.NextStepHint("feature"))
	return nil
}

func newFeatureCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "feature <name>",
		Short: "Start a new feature",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return startWork(args[0], "feature")
		},
	}
}

func newFixCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "fix <name>",
		Short: "Start a bug fix",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return startWork(args[0], "fix")
		},
	}
}

func newDocsCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "docs <name>",
		Short: "Start a documentation change",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return startWork(args[0], "docs")
		},
	}
}

func newChoreCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "chore <name>",
		Short: "Start a chore (tooling, deps, config)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return startWork(args[0], "chore")
		},
	}
}

func newStartCmd() *cobra.Command {
	return &cobra.Command{
		Use:   "start <name>",
		Short: "Start working on something (type detected later)",
		Args:  cobra.ExactArgs(1),
		RunE: func(cmd *cobra.Command, args []string) error {
			return startWork(args[0], "unknown")
		},
	}
}

File: platform/cli/src/commands/feature.ts
import { Command } from 'commander';
import chalk from 'chalk';
import * as git from '../git';
import {
  ensureDevpodDir,
  loadConfig,
  saveFeature,
  slugify,
  getCurrentFeature,
  saveUndoEntry,
  type ChangeType,
  type FeatureData,
} from '../workspace';
import {
  errorMessage,
  contextLine,
  nextStepHint,
  featureTypePrefix,
} from '../format';

function autoSaveChanges(cwd?: string): void {
  if (!git.isClean(cwd)) {
    git.stageAll(cwd);
    git.commit('WIP: auto-save before switching context', cwd);
  }
}

function startFeature(
  name: string,
  type: ChangeType,
): void {
  // Auto-save uncommitted changes on current feature
  autoSaveChanges();

  const config = loadConfig();
  const slug = slugify(name);
  const prefix = type === 'unknown' ? 'feature' : type;
  const branch = `${prefix}/${slug}`;

  // Fetch latest and create branch
  git.fetchMain();
  git.createBranch(branch, config.defaultBranch);

  // Create feature data
  const feature: FeatureData = {
    name,
    type,
    slug,
    branch,
    created: new Date().toISOString(),
    diffs: [],
    status: 'active',
  };

  ensureDevpodDir();
  saveFeature(feature);

  console.log(
    `${chalk.green('\u2713')} Started ${featureTypePrefix(type)}: ${chalk.bold(name)}`,
  );
  console.log(contextLine(feature, null));
  console.log(nextStepHint('feature'));
}

export function registerFeatureCommands(program: Command): void {
  program
    .command('feature')
    .description('Start a new feature')
    .argument('<name>', 'Feature name')
    .action(async (name: string) => {
      try {
        startFeature(name, 'feature');
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  program
    .command('fix')
    .description('Start a bug fix')
    .argument('<name>', 'Fix description')
    .action(async (name: string) => {
      try {
        startFeature(name, 'fix');
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  program
    .command('docs')
    .description('Start a documentation change')
    .argument('<name>', 'Docs description')
    .action(async (name: string) => {
      try {
        startFeature(name, 'docs');
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  program
    .command('chore')
    .description('Start a chore (tooling, deps, config)')
    .argument('<name>', 'Chore description')
    .action(async (name: string) => {
      try {
        startFeature(name, 'chore');
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });

  program
    .command('start')
    .description('Start working on something (type detected later)')
    .argument('<name>', 'Description of what you are working on')
    .action(async (name: string) => {
      try {
        startFeature(name, 'unknown');
      } catch (err) {
        console.error(errorMessage(err));
        process.exit(1);
      }
    });
}
